File: java_code/p004_hard_median_of_two_sorted_arrays.java
// class Solution
// assume no empty list

// another possible way is every time moves forward min(len1, k/2) and k is the number of elems we need to throw
class medianOfTwoSortedArrays {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if (nums1.length > nums2.length) {
            int[] tmp = nums1;
            nums1 = nums2;
            nums2 = tmp;
        }
        int m = nums1.length;
        int n = nums2.length;
        int imin = 0;
        int imax = m;
        while (true) {
            // i: [0, m]
            int i = (imin + imax) / 2;
            // j: [0, n]
            int j = (m + n + 1) / 2 - i;
            if ((i <= 0 || j >= n || nums1[i - 1] <= nums2[j]) &&
                    (j <= 0 || i >= m || nums2[j - 1] <= nums1[i])) {
                // for even
                if ((m + n) % 2 == 0) {
                    int x1, x2;
                    if (i <= 0) {
                        x1 = nums2[j - 1];
                    } else if (j <= 0) {
                        x1 = nums1[i - 1];
                    } else {
                        x1 = Math.max(nums2[j - 1], nums1[i - 1]);
                    }
                    if (i >= m) {
                        x2 = nums2[j];
                    } else if (j >= n) {
                        x2 = nums1[i];
                    } else {
                        x2 = Math.min(nums2[j], nums1[i]);
                    }
                    return (x1 + x2) / 2.0;
                }
                // for odd
                if (i <= 0) {
                    return nums2[j - 1];
                }
                if (j <= 0) {
                    return nums1[i - 1];
                }
                return Math.max(nums2[j - 1], nums1[i - 1]);
            }
            if (j > 0 && i < m && nums2[j - 1] > nums1[i]) {
                // increase i
                imin = i + 1;
            } else {
                imax = i - 1;
            }
        }
    }
}

File: python_code/p004_hard_median_of_two_sorted_arrays.py
# need be O(log (m+n))
# make two part same len (or p1 = p2+1) at first
# assume m<n and no empty list
# carefully check boundary!
# but seems sort is faster

class Solution:
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1
        m = len(nums1)
        n = len(nums2)
        imin = 0
        imax = m
        while True:
            i = (imin + imax) // 2
            # for i[0,m] -> j>=0 and j<=n
            j = (m + n + 1) // 2 - i
            if (i <= 0 or j >= n or nums1[i - 1] <= nums2[j]) and \
                    (j <= 0 or i >= m or nums2[j - 1] <= nums1[i]):
                # for even
                if (m + n) % 2 == 0:
                    if i <= 0:
                        x1 = nums2[j - 1]
                    elif j <= 0:
                        x1 = nums1[i - 1]
                    else:
                        x1 = max(nums2[j - 1], nums1[i - 1])
                    if i >= m:
                        x2 = nums2[j]
                    elif j >= n:
                        x2 = nums1[i]
                    else:
                        x2 = min(nums1[i], nums2[j])
                    return (x1 + x2) / 2
                # for odd
                if i <= 0:
                    return nums2[j - 1]
                if j <= 0:
                    return nums1[i - 1]
                return max(nums2[j - 1], nums1[i - 1])
            if j > 0 and i < m and nums2[j - 1] > nums1[i]:  # increase i
                imin = i + 1
            else:
                imax = i - 1
